src/main/java/edu/ucla/library/bucketeer/Job.java

```diff
@@ -206,68 +206,73 @@ public String getSlackHandle() {
+
+        // Check which headers already exist in the metadata
+        final boolean bucketeerStateMissing = bucketeerStateIndex == -1;
+        final boolean accessUrlMissing = accessUrlIndex == -1;
 
         // If both headers are missing, we need to expand our headers array by two
-        if (bucketeerStateIndex == -1 && accessUrlIndex == -1) {
-            final String[] newHeader = new String[metadataHeader.length + 2];
-
-            LOGGER.debug(MessageCodes.BUCKETEER_155, 2);
-
-            System.arraycopy(metadataHeader, 0, newHeader, 0, metadataHeader.length);
-            newHeader[metadataHeader.length] = Metadata.BUCKETEER_STATE;
-            newHeader[metadataHeader.length + 1] = Metadata.IIIF_ACCESS_URL;
-            bucketeerStateIndex = metadataHeader.length;
-            accessUrlIndex = metadataHeader.length + 1;
-            metadataHeader = newHeader;
-            setMetadataHeader(metadataHeader);
-        } else if (bucketeerStateIndex == -1 || accessUrlIndex == -1) {
+        if (bucketeerStateMissing && accessUrlMissing) {
+            additionalHeadersCount = 2;
+            additionalHeaders = new String[additionalHeadersCount];
+            additionalHeaders[0] = Metadata.BUCKETEER_STATE;
+            additionalHeaders[1] = Metadata.IIIF_ACCESS_URL;
+
+            bucketeerStateIndex = myMetadataHeader.length;
+            accessUrlIndex = myMetadataHeader.length + 1;
+        } else if (bucketeerStateMissing) {
             // If only one header is missing, we need to expand our headers array by one
-            final String[] newHeader = new String[metadataHeader.length + 1];
-            final int index = findHeader(metadataHeader, Metadata.BUCKETEER_STATE);
+            additionalHeadersCount = 1;
+            additionalHeaders = new String[additionalHeadersCount];
+            additionalHeaders[0] = Metadata.BUCKETEER_STATE;
+
+            bucketeerStateIndex = myMetadataHeader.length;
+        } else if (accessUrlMissing) {
+            // Expand by one
+            additionalHeadersCount = 1;
+            additionalHeaders = new String[additionalHeadersCount];
+            additionalHeaders[0] = Metadata.IIIF_ACCESS_URL;
+
+            accessUrlIndex = myMetadataHeader.length;
+        } else {
+            // No change
+            additionalHeadersCount = 0;
+            additionalHeaders = new String[additionalHeadersCount];
+        }
 
-            LOGGER.debug(MessageCodes.BUCKETEER_155, 1);
+        LOGGER.debug(MessageCodes.BUCKETEER_155, additionalHeadersCount);
+        newHeaderLength = myMetadataHeader.length + additionalHeadersCount;
 
-            System.arraycopy(metadataHeader, 0, newHeader, 0, metadataHeader.length);
+        // Update the headers if there are changes
+        if (bucketeerStateMissing || accessUrlMissing) {
+            newHeader = new String[newHeaderLength];
 
-            // If Bucketeer State was found, add the other one; else, add Bucketeer State
```

Comment:
I know rewriting this method isn't something we discussed, but it seemed like a good idea once I started looking at it some more. The result is a little DRYer (with calls to `setMetadataHeader`, `LOGGER.debug`, and `System.arraycopy` in one place), and the first if-statement now has one level instead of two.

Reply:
Looks good to me.